### PR TITLE
Add BatchNorm1D

### PR DESCRIPTION
**Pull request by KSTconsult**

### CODE REVIEW

File: src/runtime/cpu/kernels/kernels.cpp [hunk 1]
```diff
@@ -65,3 +65,55 @@ float deepworks::CPUNLLLoss(Matrix predictions, ConstVector target) {
...
+    if (is_training) {
+        auto input_mean = input.colwise().mean();
+        input_centered = input.rowwise() - input_mean;
+
+        auto input_var = input_centered.cwiseAbs2().colwise().mean();
+        std = (input_var.array() + eps).cwiseSqrt();
+
+        running_mean = running_mean * alpha + input_mean * (1 - alpha);
+        running_var = running_var * alpha + input_var * (1 - alpha);
+
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/kernels/kernels.cpp:84]: empty line

File: src/runtime/cpu/layers/cpubatchnorm.cpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    const auto& beta_grad_shape       = m_grad_beta.shape();
+    const auto& grad_output           = grad_outputs.front();
+    const auto& grad_output_shape     = grad_output.shape();
+
+    deepworks::CPUBatchNorm1DBackwardParamGrad({m_input_centered.data(), input_centered_shape[0], input_centered_shape[1]},
+                                               {m_std.data(), std_shape[0]},
+                                               {grad_output.data(), grad_output_shape[0], grad_output_shape[1]},
+                                               {m_grad_gamma.data(), gamma_grad_shape[0]},
+                                               {m_grad_beta.data(), beta_grad_shape[0]});
+}
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/layers/cpubatchnorm.cpp:110]: New line

**Review by TolyaTalamanov — APPROVED** [review]
LTGM :+1: 

@l-bat Have a look, please. I'm OK with that if you're too :)

File: src/runtime/cpu/layers/cpubatchnorm.cpp [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+deepworks::cpu::CPUBatchNorm1D::CPUBatchNorm1D(deepworks::LayerInfo&& info)
+    : deepworks::cpu::ICPULayer(std::move(info)),
+      m_gamma(m_info.params()[0].data()),
+      m_beta(m_info.params()[1].data()),
+      m_grad_gamma(m_info.params()[0].grad()),
+      m_grad_beta(m_info.params()[1].grad()) {
+
+    m_std.allocate({m_gamma.shape()[0]});
+    m_running_mean.allocate({m_gamma.shape()[0]});
+    m_running_var.allocate({m_gamma.shape()[0]});
```
Comment (l-bat) [inline src/runtime/cpu/layers/cpubatchnorm.cpp:14]: I'm not sure, that we need to allocate internal tensors in constructor.

File: src/runtime/cpu/kernels/kernels.hpp [hunk 4]
```diff
@@ -112,4 +112,47 @@ std::vector<int> MatchTargetTo1dMatrix(ConstVector target, int batch_size, int n
  */
 float CPUNLLLoss(Matrix predictions, ConstVector target);
 
+/*
+ * CPUBatchNormForward
```
Comment (l-bat) [inline src/runtime/cpu/kernels/kernels.hpp:116]:
```suggestion
 * CPUBatchNorm1DForward
```

File: src/runtime/cpu/kernels/kernels.hpp [hunk 5]
```diff
@@ -112,4 +112,47 @@ std::vector<int> MatchTargetTo1dMatrix(ConstVector target, int batch_size, int n
...
+/*
+ * CPUBatchNormBackward
+ * Implements Batch Normalization backward pass for input
+ * input_centered have size [batch_size, in_features]
+ * std have size [1, in_features]
+ * grad_output have size [batch_size, in_features]
+ * grad_input have size [batch_size, in_features]
+ * gamma have size [1, in_features]
+ */
+void CPUBatchNorm1DBackwardInputGrad(ConstMatrix input_centered, ConstVector std,
```
Comment (l-bat) [inline src/runtime/cpu/kernels/kernels.hpp:142]: Let's rename it to `CPUBatchNorm1DInputGrad`  for consistency with other kernel names.

File: src/runtime/cpu/kernels/kernels.hpp [hunk 6]
```diff
@@ -112,4 +112,47 @@ std::vector<int> MatchTargetTo1dMatrix(ConstVector target, int batch_size, int n
...
+/*
+ * CPUBatchNormBackward
+ * Implements Batch Normalization backward pass for params
+ * input_centered have size [batch_size, in_features]
+ * std have size [1, in_features]
+ * grad_output have size [batch_size, in_features]
+ * gamma_grad have size [1, in_features]
+ * beta_grad have size [1, in_features]
+ */
+void CPUBatchNorm1DBackwardParamGrad(ConstMatrix input_centered, ConstVector std, ConstMatrix grad_output,
```
Comment (l-bat) [inline src/runtime/cpu/kernels/kernels.hpp:155]: Let's rename it to `CPUBatchNorm1DParamGrad` and update docs

File: src/runtime/cpu/layers/cpubatchnorm.cpp (hunk 3 above)
Comment (l-bat) [reply src/runtime/cpu/layers/cpubatchnorm.cpp:14]: OK, let's move it to `init` later

**Review by l-bat — APPROVED** [review]

**Review by l-bat — APPROVED** [review]

File: src/expression/nn.cpp [hunk 7]
```diff
@@ -45,3 +45,26 @@ deepworks::Softmax::Softmax(std::string name)
...
 }
+
+deepworks::BatchNorm1D::BatchNorm1D(float eps, float alpha, std::string name)
+    : BaseOp<deepworks::BatchNorm1D>(LayerInfo(std::move(name), "BatchNorm1D")) {
+    m_info.impl().attrs["eps"] = eps;
+    m_info.impl().attrs["alpha"] = alpha;
+}
+
+void deepworks::BatchNorm1D::init(const Shape& in_shape) {
+
```
Comment (TolyaTalamanov) [inline src/expression/nn.cpp:56]: empty line

File: src/runtime/cpu/kernels/kernels.cpp [hunk 8]
```diff
@@ -65,3 +65,54 @@ float deepworks::CPUNLLLoss(Matrix predictions, ConstVector target) {
...
+        std = (running_var.array() + eps).cwiseSqrt();
+    }
+
+    output = (input_centered.array().rowwise() * (gamma.array() / std.array())).array().rowwise() + beta.array();
+}
+
+void deepworks::CPUBatchNorm1DInputGrad(ConstMatrix input_centered, ConstVector std,
+                                        ConstMatrix grad_output, Matrix grad_input,
+                                        ConstVector gamma) {
+
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/kernels/kernels.cpp:95]: empty line

File: src/runtime/cpu/layers/cpubatchnorm.cpp [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    // If change batch_size need to create new m_input_centered tensor
+    if (m_input_centered.shape() != input.shape()) {
+        m_input_centered = Tensor(in_shape);
+    }
+
+    // Check if model in train mode
+    bool gamma_is_trainable = m_info.params()[0].is_trainable();
+    bool beta_is_trainable = m_info.params()[1].is_trainable();
+
+    bool is_trainable = gamma_is_trainable == beta_is_trainable;
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/layers/cpubatchnorm.cpp:49]: so, what if `gamma_is_trainable` == false && `betas_is_trainable` == false. is_trainable == true ?

**Comment by TolyaTalamanov** [comment]
Please reply on comments if you've already fixed it

File: src/runtime/cpu/kernels/kernels.cpp (hunk 1 above)
Comment (KSTconsult) [reply src/runtime/cpu/kernels/kernels.cpp]: Done.

File: src/runtime/cpu/layers/cpubatchnorm.cpp (hunk 2 above)
Comment (KSTconsult) [reply src/runtime/cpu/layers/cpubatchnorm.cpp]: Done.

File: src/runtime/cpu/kernels/kernels.hpp (hunk 4 above)
Comment (KSTconsult) [reply src/runtime/cpu/kernels/kernels.hpp]: Done.

File: src/runtime/cpu/kernels/kernels.hpp (hunk 5 above)
Comment (KSTconsult) [reply src/runtime/cpu/kernels/kernels.hpp]: Done.

File: src/runtime/cpu/kernels/kernels.hpp (hunk 6 above)
Comment (KSTconsult) [reply src/runtime/cpu/kernels/kernels.hpp]: Done.

File: src/expression/nn.cpp (hunk 7 above)
Comment (KSTconsult) [reply src/expression/nn.cpp]: Done.

File: src/runtime/cpu/kernels/kernels.cpp (hunk 8 above)
Comment (KSTconsult) [reply src/runtime/cpu/kernels/kernels.cpp]: Done.

File: src/runtime/cpu/layers/cpubatchnorm.cpp (hunk 9 above)
Comment (KSTconsult) [reply src/runtime/cpu/layers/cpubatchnorm.cpp]: Done.

File: src/runtime/cpu/layers/cpubatchnorm.cpp [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    // If change batch_size need to create new m_input_centered tensor
+    if (m_input_centered.shape() != input.shape()) {
+        m_input_centered = Tensor(in_shape);
+    }
+
+    // Check if model in train mode
+    bool gamma_is_trainable = m_info.params()[0].is_trainable();
+    bool beta_is_trainable = m_info.params()[1].is_trainable();
+
+    bool is_trainable = gamma_is_trainable && beta_is_trainable;
```
Comment (TolyaTalamanov) [inline src/runtime/cpu/layers/cpubatchnorm.cpp:49]: In fact I'm expecting something like this: 
```
DeepWorks_Assert(gamma_is_trainable == beta_is_trainable &&
                                "Gamma and beta should have same trainable mode");
bool is_trainable = gamma_is_trainable;
```

but, let it be

**Comment by TolyaTalamanov** [comment]
Good job, Thanks! :+1:

**Pull request merged**